scheduling/dispatcher.rb

```diff
@@ -1,93 +1,228 @@
-        Kernel.exit!
-
-        # rubocop:disable Lint/UnreachableCode
-        # Reachable in test only.
-        next
-        # rubocop:enable Lint/UnreachableCode
-      end
-
-      ready.first.close
-    end.tap { _1.name = "apoptosis:" + strand_ubid }
-
-    Thread.new do
-      Thread.current[:created_at] = Time.now
-      strand.run Strand::LEASE_EXPIRATION / 4
-    rescue => ex
-      Clog.emit("exception terminates thread") { Util.exception_to_hash(ex) }
-
-      loop do
-        ex = ex.cause
-        break unless ex
-        Clog.emit("nested exception") { Util.exception_to_hash(ex) }
-      end
-    ensure
-      # Adequate to unblock IO.select.
-      apoptosis_w.close
-      notify_w.close
-    end.tap { _1.name = strand_ubid }
-
-    notify_r
+  # Handle timeout of a strand thread by killing the process.
+  def apoptosis_failure
+    # Timed out, dump threads and exit.
+    # Don't thread print concurrently.
+    APOPTOSIS_MUTEX.synchronize do
+      ThreadPrinter.run
+      Kernel.exit!
+    end
   end
 
-  def start_cohort
-    scan.each do |strand|
-      break if shutting_down
-      @notifiers << start_strand(strand)
+  # The entry point for strand threads.  The loops until the
+  # dispatcher shuts down, monitoring the strand queue, signalling
+  # the related apoptosis thread when starting, running the strand,
+  # and then signalling the related apoptosis thread when the
+  # strand run finishes.
+  def strand_thread(strand_queue, start_queue, finish_queue)
+    run_strand(strand_queue.pop, start_queue, finish_queue) until @shutting_down
+  ensure
+    # Signal related apoptosis thread to shutdown
+    start_queue.push(nil)
+  end
+
+  # Handle the running of a single strand.  Signals the apoptosis
+  # thread via the start queue when starting, and the finish queue
+  # when exiting.
+  def run_strand(strand, start_queue, finish_queue)
+    # Shutdown indicated, return immediately
+    return unless strand
+
+    strand_ubid = strand.ubid.freeze
+    Thread.current.name = strand_ubid
+    start_queue.push(strand_ubid)
+    strand.run(STRAND_RUNTIME)
+  rescue => ex
+    Clog.emit("exception terminates strand run") { Util.exception_to_hash(ex) }
+
+    cause = ex
+    loop do
+      break unless (cause = cause.cause)
+      Clog.emit("nested exception") { Util.exception_to_hash(cause) }
     end
+    ex
+  ensure
+    # Always signal apoptosis thread that the strand has finished,
+    # even for non-StandardError exits
+    finish_queue.push(true)
+    @mutex.synchronize { @current_strands.delete(strand&.id) }
   end
 
-  def wait_cohort
-    if shutting_down
-      Clog.emit("Shutting down.") { {unfinished_strand_count: @notifiers.length} }
-      Kernel.exit if @notifiers.empty?
+  # Find strands that need to be run, and push each onto the
+  # strand queue.  This can block if the strand queue is full,
+  # to allow for backoff in the case of a busy thread pool.
+  def start_cohort
+    strand_queue = @strand_queue
```

Comment:
Let's say we have this situation (schedule and strand id):
00:00:00 strand 1
NOW
00:00:01 strand 2
00:00:01 strand 3
...
00:00:01 strand 100

scan only returns "strand 1" but within a second many more strands are due. "strand 1" runs for one minute. will strands 2-100 be executed at ~00:00:01  or at ~00:01:00?

Reply:
strands 2-100 would be enqueued to the job pool around `~00:00:01` (though enqueing may block if it exceeds the queue buffer size).  When strands start processing depends on thread pool size.  With a thread pool of size 1, strand 2 would start at `~00:01:00`, with others following that.  With a thread pool size of 2 or more, strand 2 would start around `~00:00:01`.  I'm guessing we use a decent number of threads for the production thread pool.